src/error.jl

```diff
@@ -27,16 +25,62 @@ end
 function gdaljl_errorhandler(class::CPLErr, errno::Cint, errmsg::Cstring)
     # function signature needs to match the one in __init__, and the signature
     # of the callback for a custom error handler in the GDAL docs
-    if class in throw_class
+    if class === CE_Failure
         throw(GDALError(class, errno, unsafe_string(errmsg)))
     end
     return C_NULL
 end
 
-"Throw an error if a pointer is null and GDAL reports an error"
-function failsafe(ptr)
-    if ptr == C_NULL && cplgetlasterrortype() in throw_class
+"Check the last error type and throw a GDALError if it is a failure"
+function maybe_throw()
+    if cplgetlasterrortype() === CE_Failure
         throw(GDALError())
     end
-    ptr
+    nothing
+end
+
+"""
+Handle anything returned from GDAL
+
+When values are returned from GDAL, always check if there was a failure using `maybe_throw`.
+
+Depending on the return type, we do extra work.
+"""
+function aftercare(x)
+    maybe_throw()
+    x
+end
+
+"For string pointers, load them to String, and free them if we should."
+function aftercare(ptr::Cstring, free::Bool)
+    maybe_throw()
+    if ptr == C_NULL
+        return nothing
+    else
+        s = unsafe_string(ptr)
+        free && vsifree(convert(Ptr{Cvoid}, ptr))
+        return s
+    end
+end
+
+"""
+For string list pointers, load them to Vector{String}
+
+That is it expects a `StringList`, in the sense of the CPL functions,
+as a null-terminated array of strings.
+"""
+function aftercare(ptr::Ptr{Cstring})
```

Comment:
Do these pointers never need to be freed like the string pointer above?

Reply:
Yeah they do at least sometimes, see for instance:

`CPLGetConfigOptions`: a copy of the list, to be freed with `CSLDestroy()`

I didn't add it here yet since it is an improvement that we can add at any time, and because that function lives in `cpl_string.h` which we don't wrap. Which doesn't mean of course we cannot just directly ccall the function. Would be good to make an issue as a reminder though.

Reply:
Ok, we can create an issue of it. Maybe add it as a `# todo` in the code?

Reply:
Yep! See 259a27c8ce78e53a078dc624cf5c3bec9a5b07f6 and #77.